src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -130,7 +109,39 @@ public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle,
             return result;
         }
 
-        private static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        // Returns the first header or null if header isn't found.
+        public static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        {
+            uint index = 0;
+            
+            return GetResponseHeaderStringHelper(requestHandle, infoLevel, ref index);
+        }
+
+        // Returns all headers or an empty list if header isn't found.
+        public static List<string> GetResponseHeaders(SafeWinHttpHandle requestHandle, uint infoLevel)
+        {
+            uint index = 0;
+            var headerList = new List<string>();
```

Comment:
You could only allocate the list when there is a header to add, and either return null or an empty list for no headers. I'm not a huge fan of returning null, but it might be easiest here and there are other methods that already return null for no headers, e.g. `GetResponseHeaderStringInfo` when the header isn't found (callers would have to check for null before foreaching over the list). Returning a static readonly empty list would also work as the list is only used internally, so we know nobody is going to add to it (and asserts could be added to ensure it stays empty).

Something like:

```c#
public static List<string> GetResponseHeaders(SafeWinHttpHandle requestHandle, uint infoLevel)
{
    uint index = 0;
    List<string> headerList = null;
    string header;
    
    while (true)
    {
        header = GetResponseHeaderStringHelper(requestHandle, infoLevel, ref index);

        if (header != null)
        {
            if (headerList == null)
            {
                headerList = new List<string>();
            }

            headerList.Add(header);
        }
        else
        {
            break;
        }
    }
    
    return headerList;
}
```